### ISSUE TITLE
Implement java.util.Formatter 

### ISSUE THREAD

**Issue by densh** [scala-native/scala-native#245]
Can be ported from Harmony.

**Comment by soc** [comment]
Probably coordinate with @cquiroz, as he is working on a lot of related stuff.

**Comment by marlomajor** [comment]
I can work on this. Can you describe the problem to me? I'm new to Scala so please be as specific as possible.

**Comment by MasseGuillaume** [comment]
Hi @marlomajor I'm new to scala-native but I have some experience in Scala. Ping me if you need help.

**Comment by MasseGuillaume** [comment]
@densh I need a test framework for this task, I'm starting from https://github.com/scala-native/scala-native/pull/263

@marlomajor If you want to join me in this PR I would be really happy to do some pair programming.

**Comment by MasseGuillaume** [comment]
scala-js implementation: https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/Formatter.scala#L1

**Comment by marlomajor** [comment]
Hey I just saw this comment I apologize. Yes, I would like to work on this with you.

**Comment by marlomajor** [comment]
I would like to pair program on this if the chance is still available

**Comment by marlomajor** [comment]
How similar to scala.js will the java.util.Formatter be for scala-native?

**Comment by sjrd** [comment]
Very similar. The main difference is that Scala.js uses `js.RegExp` to "parse" the format string, which SN will have to do in a different way, obviously. But everything else should be copy-paste.

**Comment by marlomajor** [comment]
Okay. What is the high-level goal behind this implementation?

**Comment by marlomajor** [comment]
Will submit pr soon.

**Comment by marlomajor** [comment]
Is there a test suite I can follow?

**Comment by sjrd** [comment]
https://github.com/scala-js/scala-js/blob/master/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala

**Comment by densh** [comment]
Indeed, missing support for regular expressions is the main hindrance to fixing this.

**Comment by sjrd** [comment]
I don't think regex support is necessary to implement `Formatter`. It is perfectly reasonable to write a parser for `%` formats by hand. In fact, it is probably a good idea to do so even if regex are supported for 1) performance and 2) not bringing the regex engine into every app that uses `f""`

**Comment by MasseGuillaume** [comment]
@marlomajor I'm on the East Coast what about next week end? Are you available?

**Comment by MasseGuillaume** [comment]
@densh what's the context for this issue. Why you need java.util.Formatter in particular ?

**Comment by MasseGuillaume** [comment]
I started a branch here: https://github.com/MasseGuillaume/scala-native/commit/6b6bd1247c869cbef38de2ad2439108de6963048

@densh It's a bit slow to to execute `sbt run` ~40 sec is this normal ?

**Comment by densh** [comment]
Linking is quite slow at the moment, I'm working on making it faster in https://github.com/scala-native/scala-native/tree/topic/fat-code-gen.

**Comment by densh** [comment]
@MasseGuillaume We need `java.util.Formatter` to support http://docs.scala-lang.org/overviews/core/string-interpolation.html and friends.

**Comment by MasseGuillaume** [comment]
@densh why not using `fprintf`?

**Comment by sjrd** [comment]
In Scala.js, we "cheated" a bit for `Formatter`. We assumed en-US in the implementation of `Formatter`. It was at a time when we did not have such strong requirements for implementations of the JDK. Nowadays, the golden rule is: if it links, it must honor its specification. We can still sort of argue that `Formatter` honors its specification based on the fact that setting the default locale does not link, and we can prevent the default default locale (double-default intended) is always en-US. But it's still cheating, if we're not lying to ourselves.

**Comment by MasseGuillaume** [comment]
@sjrd 
What would be the specification?

* [Oracle's JDK documentation](https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html) (Clean room)
* [The JCK/TCK](http://openjdk.java.net/groups/conformance/JckAccess/) (GPL License)
* Something else ?

@densh 
I suggest we scope this ticket to the [The f Interpolator]( 
http://docs.scala-lang.org/overviews/core/string-interpolation.html#the-f-interpolator)

In other words, only what is accepted in the f interpolator would be implemented. We would also need to implement java.util.Locale (#247).

This means I would not implement constructors related to reading from files. 
[public Formatter(File file) throws FileNotFoundException](https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html#Formatter-java.io.File-)

**Comment by sjrd** [comment]
It is off course the JavaDoc (clean room). In addition, when the JavaDoc is imprecise or downright incorrect, we reverse-engineer a more precise specification by experimenting with the observable behavior of the JVM JDK (typically through the Scala REPL).

We turn all that into a test suite of our own, which we cross-compile for JVM and JS. The JS version tests or implementation. The JVM version tests that the test suite is correct (i.e., it characterizes what happens on the JVM).

**Comment by densh** [comment]
@MasseGuillaume Scoping this issue to `f interpolator` is a good idea. That's the main use case at the moment.

**Comment by MasseGuillaume** [comment]
@densh cool. I need to focus on Scastie / moving to Lausanne. I will be back on this issue soon ;-)

**Comment by densh** [comment]
There was an attempt by @hardmettle to fix this in #529 but he didn't have time to finish it. It's up for grabs if anyone is still interested.

**Comment by EDawg878** [comment]
@densh I'm interested in working on this. Just to be clear, it seems like most of the code can be directly taken from the Scala.Js implementation but we the caveat is that we have to manually parse instead of using regex? If that's the case, then I'll see what I can do.

**Comment by MasseGuillaume** [comment]
@EDawg878 is it ok if I take this ticket?

**Comment by olafurpg** [comment]
I hit on ??? in Formatter

```

scala.NotImplementedError: an implementation is missing
	at java.lang.Throwable::init_java.lang.String_java.lang.Throwable
	at java.lang.Error::init_java.lang.String_java.lang.Throwable
	at java.lang.Error::init_java.lang.String
	at scala.NotImplementedError::init_java.lang.String
	at scala.NotImplementedError::init
	at scala.Predef$::???_nothing
	at java.util.Formatter::format_java.lang.String_scala.scalanative.runtime.ObjectArray_java.util.Formatter
	at java.lang.String$::format_java.util.Locale_java.lang.String_scala.scalanative.runtime.ObjectArray_java.lang.String
	at java.lang.String$::format_java.lang.String_scala.scalanative.runtime.ObjectArray_java.lang.String
	at scala.collection.immutable.StringLike$class$::format_scala.collection.immutable.StringLike_scala.collection.Seq_java.lang.String
	at scala.collection.immutable.StringOps::format_scala.collection.Seq_java.lang.String
	at metaconfig.ConfOps$::show_metaconfig.Conf_java.lang.String
```

while running scalafmt on native.

**Comment by densh** [comment]
Fixed in #804
